File: src/actions/change-requests.ts
"use server";

import { prisma } from "@/lib/db";
import { createAuditLog } from "@/lib/audit";
import {
  ChangeRequestFormSchema,
  type ChangeRequestFormData,
} from "@/lib/validations/change-request";
import { ChangeRequestStatus } from "@prisma/client";
import { logger } from "@/lib/logger";

export async function getChangeRequests(projectId: string) {
  try {
    const changeRequests = await prisma.changeRequest.findMany({
      where: { projectId },
      orderBy: { createdAt: "desc" },
      include: {
        createdBy: {
          select: { id: true, name: true },
        },
        assignedTo: {
          select: { id: true, name: true },
        },
      },
    });

    return { success: true, changeRequests };
  } catch (error) {
    logger.error("getChangeRequests error:", error);
    return { success: false, error: "Failed to fetch change requests" };
  }
}

export async function getChangeRequest(id: string) {
  try {
    const changeRequest = await prisma.changeRequest.findUnique({
      where: { id },
      include: {
        project: {
          select: { id: true, name: true, slug: true },
        },
        createdBy: {
          select: { id: true, name: true, email: true },
        },
        assignedTo: {
          select: { id: true, name: true, email: true },
        },
      },
    });

    if (!changeRequest) {
      return { success: false, error: "Change request not found" };
    }

    return { success: true, changeRequest };
  } catch (error) {
    logger.error("getChangeRequest error:", error);
    return { success: false, error: "Failed to fetch change request" };
  }
}

export async function createChangeRequest(
  data: ChangeRequestFormData,
  actorUserId: string
) {
  try {
    const validated = ChangeRequestFormSchema.parse(data);

    const changeRequest = await prisma.changeRequest.create({
      data: {
        projectId: validated.projectId,
        createdByUserId: actorUserId,
        title: validated.title,
        description: validated.description,
        sourceType: validated.sourceType,
        status: validated.status,
        impact: validated.impact,
        githubIssueUrl: validated.githubIssueUrl || null,
        githubBranch: validated.githubBranch ?? null,
        githubPrUrl: validated.githubPrUrl || null,
        assignedToUserId: validated.assignedToUserId ?? null,
      },
    });

    await createAuditLog({
      actorUserId,
      entityType: "ChangeRequest",
      entityId: changeRequest.id,
      action: "CREATE",
      metadata: {
        projectId: validated.projectId,
        title: validated.title,
        impact: validated.impact,
      },
    });

    return { success: true, changeRequest };
  } catch (error) {
    logger.error("createChangeRequest error:", error);
    return { success: false, error: "Failed to create change request" };
  }
}

export async function updateChangeRequest(
  id: string,
  data: Partial<ChangeRequestFormData & { status: ChangeRequestStatus }>,
  actorUserId: string
) {
  try {
    const existing = await prisma.changeRequest.findUnique({ where: { id } });
    if (!existing) {
      return { success: false, error: "Change request not found" };
    }

    const changeRequest = await prisma.changeRequest.update({
      where: { id },
      data: {
        ...(data.title !== undefined ? { title: data.title } : {}),
        ...(data.description !== undefined ? { description: data.description } : {}),
        ...(data.sourceType !== undefined ? { sourceType: data.sourceType } : {}),
        ...(data.status !== undefined ? { status: data.status } : {}),
        ...(data.impact !== undefined ? { impact: data.impact } : {}),
        ...(data.githubIssueUrl !== undefined
          ? { githubIssueUrl: data.githubIssueUrl || null }
          : {}),
        ...(data.githubBranch !== undefined
          ? { githubBranch: data.githubBranch ?? null }
          : {}),
        ...(data.githubPrUrl !== undefined
          ? { githubPrUrl: data.githubPrUrl || null }
          : {}),
        ...(data.assignedToUserId !== undefined
          ? { assignedToUserId: data.assignedToUserId ?? null }
          : {}),
      },
    });

    if (data.status && data.status !== existing.status) {
      await createAuditLog({
        actorUserId,
        entityType: "ChangeRequest",
        entityId: id,
        action: "STATUS_CHANGE",
        metadata: {
          from: existing.status,
          to: data.status,
          title: changeRequest.title,
        },
      });
    } else {
      await createAuditLog({
        actorUserId,
        entityType: "ChangeRequest",
        entityId: id,
        action: "UPDATE",
        metadata: { title: changeRequest.title },
      });
    }

    return { success: true, changeRequest };
  } catch (error) {
    logger.error("updateChangeRequest error:", error);
    return { success: false, error: "Failed to update change request" };
  }
}

export async function reopenChangeRequest(id: string, actorUserId: string) {
  try {
    const existing = await prisma.changeRequest.findUnique({ where: { id } });
    if (!existing) {
      return { success: false, error: "Change request not found" };
    }

    const changeRequest = await prisma.changeRequest.update({
      where: { id },
      data: {
        status: ChangeRequestStatus.NEW,
        reopenedCount: { increment: 1 },
        closedAt: null,
      },
    });

    await createAuditLog({
      actorUserId,
      entityType: "ChangeRequest",
      entityId: id,
      action: "REOPEN",
      metadata: {
        previousStatus: existing.status,
        reopenedCount: changeRequest.reopenedCount,
        title: changeRequest.title,
      },
    });

    return { success: true, changeRequest };
  } catch (error) {
    logger.error("reopenChangeRequest error:", error);
    return { success: false, error: "Failed to reopen change request" };
  }
}

export async function closeChangeRequest(id: string, actorUserId: string) {
  try {
    const existing = await prisma.changeRequest.findUnique({ where: { id } });
    if (!existing) {
      return { success: false, error: "Change request not found" };
    }

    const changeRequest = await prisma.changeRequest.update({
      where: { id },
      data: {
        status: ChangeRequestStatus.DONE,
        closedAt: new Date(),
      },
    });

    await createAuditLog({
      actorUserId,
      entityType: "ChangeRequest",
      entityId: id,
      action: "CLOSE",
      metadata: {
        previousStatus: existing.status,
        title: changeRequest.title,
      },
    });

    return { success: true, changeRequest };
  } catch (error) {
    logger.error("closeChangeRequest error:", error);
    return { success: false, error: "Failed to close change request" };
  }
}

File: backend/app/routers/change_requests.py
from fastapi import APIRouter, Depends, HTTPException, status, Request
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from datetime import datetime, timezone

from app.database import get_db
from app.core.dependencies import require_role, get_client_ip
from app.core.rbac import Role
from app.core.security import sanitize_html
from app.services.audit_service import create_audit_log
from app.models.change_request import ChangeRequest, ChangeRequestStatus
from app.models.project import ProjectWorkspace
from app.schemas.change_request import ChangeRequestCreate, ChangeRequestUpdate, ChangeRequestResponse

router = APIRouter(prefix="/api/v1", tags=["change-requests"])


@router.get("/projects/{project_id}/change-requests", response_model=list[ChangeRequestResponse])
async def list_change_requests(
    project_id: str,
    current_user=Depends(require_role(Role.ADMIN, Role.EMPLOYEE)),
    db: AsyncSession = Depends(get_db),
) -> list[ChangeRequestResponse]:
    project = await db.execute(
        select(ProjectWorkspace).where(ProjectWorkspace.id == project_id, ProjectWorkspace.deleted_at.is_(None))
    )
    if not project.scalar_one_or_none():
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found")

    result = await db.execute(
        select(ChangeRequest)
        .where(ChangeRequest.project_id == project_id)
        .order_by(ChangeRequest.created_at.desc())
    )
    return [ChangeRequestResponse.model_validate(cr) for cr in result.scalars().all()]


@router.post("/projects/{project_id}/change-requests", response_model=ChangeRequestResponse, status_code=status.HTTP_201_CREATED)
async def create_change_request(
    project_id: str,
    body: ChangeRequestCreate,
    request: Request,
    current_user=Depends(require_role(Role.ADMIN, Role.EMPLOYEE)),
    db: AsyncSession = Depends(get_db),
) -> ChangeRequestResponse:
    project = await db.execute(
        select(ProjectWorkspace).where(ProjectWorkspace.id == project_id, ProjectWorkspace.deleted_at.is_(None))
    )
    if not project.scalar_one_or_none():
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found")

    cr = ChangeRequest(
        project_id=project_id,
        title=body.title,
        description=sanitize_html(body.description),
        source_type=body.source_type.value,
        status=body.status.value,
        impact=body.impact.value,
        github_issue_url=body.github_issue_url,
        github_branch=body.github_branch,
        github_pr_url=body.github_pr_url,
        created_by_user_id=current_user.id,
        assigned_to_user_id=body.assigned_to_user_id,
    )
    db.add(cr)
    await db.flush()

    await create_audit_log(
        db, "ChangeRequest", cr.id, "CREATE",
        actor_user_id=current_user.id,
        metadata={"project_id": project_id, "title": cr.title},
        ip_address=get_client_ip(request),
    )
    return ChangeRequestResponse.model_validate(cr)


@router.get("/change-requests/{cr_id}", response_model=ChangeRequestResponse)
async def get_change_request(
    cr_id: str,
    current_user=Depends(require_role(Role.ADMIN, Role.EMPLOYEE)),
    db: AsyncSession = Depends(get_db),
) -> ChangeRequestResponse:
    result = await db.execute(select(ChangeRequest).where(ChangeRequest.id == cr_id))
    cr = result.scalar_one_or_none()
    if not cr:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Change request not found")
    return ChangeRequestResponse.model_validate(cr)


@router.patch("/change-requests/{cr_id}", response_model=ChangeRequestResponse)
async def update_change_request(
    cr_id: str,
    body: ChangeRequestUpdate,
    request: Request,
    current_user=Depends(require_role(Role.ADMIN, Role.EMPLOYEE)),
    db: AsyncSession = Depends(get_db),
) -> ChangeRequestResponse:
    result = await db.execute(select(ChangeRequest).where(ChangeRequest.id == cr_id))
    cr = result.scalar_one_or_none()
    if not cr:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Change request not found")

    changes = {}
    update_data = body.model_dump(exclude_unset=True)

    # Map of allowed updatable fields
    updatable_fields = {
        "title", "description", "status", "impact",
        "github_issue_url", "github_branch", "github_pr_url",
        "assigned_to_user_id",
    }

    for field, value in update_data.items():
        if field not in updatable_fields:
            continue
        if field == "description" and value:
            value = sanitize_html(value)
        if hasattr(value, "value"):
            value = value.value
        old_value = getattr(cr, field)
        if old_value != value:
            changes[field] = {"old": str(old_value), "new": str(value)}
            if field == "title":
                cr.title = value
            elif field == "description":
                cr.description = value
            elif field == "status":
                cr.status = value
            elif field == "impact":
                cr.impact = value
            elif field == "github_issue_url":
                cr.github_issue_url = value
            elif field == "github_branch":
                cr.github_branch = value
            elif field == "github_pr_url":
                cr.github_pr_url = value
            elif field == "assigned_to_user_id":
                cr.assigned_to_user_id = value

    if changes:
        await db.flush()
        await db.refresh(cr)
        action = "STATUS_CHANGE" if "status" in changes else "UPDATE"
        await create_audit_log(
            db, "ChangeRequest", cr.id, action,
            actor_user_id=current_user.id,
            metadata=changes,
            ip_address=get_client_ip(request),
        )
    return ChangeRequestResponse.model_validate(cr)


@router.post("/change-requests/{cr_id}/reopen", response_model=ChangeRequestResponse)
async def reopen_change_request(
    cr_id: str,
    request: Request,
    current_user=Depends(require_role(Role.ADMIN, Role.EMPLOYEE)),
    db: AsyncSession = Depends(get_db),
) -> ChangeRequestResponse:
    result = await db.execute(select(ChangeRequest).where(ChangeRequest.id == cr_id))
    cr = result.scalar_one_or_none()
    if not cr:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Change request not found")

    old_status = cr.status
    cr.status = ChangeRequestStatus.NEW.value
    cr.reopened_count += 1
    cr.closed_at = None
    db.add(cr)

    await create_audit_log(
        db, "ChangeRequest", cr.id, "REOPEN",
        actor_user_id=current_user.id,
        metadata={"old_status": old_status, "reopened_count": cr.reopened_count},
        ip_address=get_client_ip(request),
    )
    return ChangeRequestResponse.model_validate(cr)


@router.post("/change-requests/{cr_id}/close", response_model=ChangeRequestResponse)
async def close_change_request(
    cr_id: str,
    request: Request,
    current_user=Depends(require_role(Role.ADMIN, Role.EMPLOYEE)),
    db: AsyncSession = Depends(get_db),
) -> ChangeRequestResponse:
    result = await db.execute(select(ChangeRequest).where(ChangeRequest.id == cr_id))
    cr = result.scalar_one_or_none()
    if not cr:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Change request not found")

    old_status = cr.status
    cr.status = ChangeRequestStatus.DONE.value
    cr.closed_at = datetime.now(timezone.utc)
    db.add(cr)

    await create_audit_log(
        db, "ChangeRequest", cr.id, "CLOSE",
        actor_user_id=current_user.id,
        metadata={"old_status": old_status},
        ip_address=get_client_ip(request),
    )
    return ChangeRequestResponse.model_validate(cr)
